### PR TITLE
Update and prepare for python3

### PR DESCRIPTION
**Pull request by eugenkleinschmidt**
Update to python3 and state-of-art python dev and structure:
- src structure
- black, ruff for linting and auto-formatting
- mypy type checking
- pytest, tox for testing

I hope it's not to many changes 😅

### CODE REVIEW

File: src/relaycard/card.py [hunk 1]
```diff
@@ -1,38 +1,47 @@
...
+    RESP_GETPORT,
+    RESP_SETPORT,
+    RESP_SETSINGLE,
+    RESP_TOGGLE,
+)
+from .frame import RequestFrame, ResponseFrame
+from .state import RelayState
 
-class RelayCard(object):
-    port = None
```
Comment (stephrdev) [inline src/relaycard/card.py:13]: Is there a reason why you removed the port and card_count initial values? Without these, the `is_initialized` property would explode if called before `setup`.

File: dev_requirements.txt [hunk 2]
```diff
@@ -0,0 +1,5 @@
+black
```
Comment (stephrdev) [inline dev_requirements.txt:1]: Have you considered using poetry or rye to manage dependencies? As you also introduced ruff (which I like!), we might also consider having a nice tool for the dependencies in place. :)

File: setup.py [hunk 3]
```diff
@@ -1,28 +1,35 @@
...
+from setuptools import find_packages, setup
+
 
 def read(fname):
     return codecs.open(os.path.join(os.path.dirname(__file__), fname)).read()
 
+
 setup(
-    name='relaycard',
-    version='0.1',
```
Comment (stephrdev) [inline setup.py:9]: As soon as we merge this, we should bump the version to 1.0 as we introduce a major breaking change by dropping Python 2 (finally).

File: src/relaycard/card.py [hunk 4]
```diff
@@ -47,70 +56,70 @@ def execute(self, command, address=0, data=0):
...
     def execute_retry(self, command, validator, address=0, data=0, retries=3):
-        for i in range(1, retries+1):
+        for i in range(1, retries + 1):
             try:
                 response = self.execute(command, address, data)
                 assert validator(response)
                 return response
-            except AssertionError, e:
-                logging.error('Error, retry #%d: %s' % (i, e))
+            except AssertionError as e:
```
Comment (stephrdev) [inline src/relaycard/card.py:64]: This is quite old code and not best practice (which I only learned over the years). If you are motivated we should get rid of all the asserts as they are not meant to be used in production code. we might simply introduce a `RelayCardException` to make it not to overly complicated but not using `assert` is a think to consider. (see https://snyk.io/blog/the-dangers-of-assert-in-python/ for more details)

File: src/relaycard/frame.py [hunk 5]
```diff
@@ -49,5 +48,9 @@ def __init__(self, response):
         self.crc = response[3]
 
     def __repr__(self):
-        return '<ResponseFrame %s/addr:%s data:%s crc:%s>' % (
-            RESPONSES[self.command], self.address, self.data, self.crc)
+        return "<ResponseFrame {}/addr:{} data:{} crc:{}>".format(
```
Comment (stephrdev) [inline src/relaycard/frame.py:51]: Have you considered migrating to f-strings right away if we touch this piece of code anyway?

File: tox.ini [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+deps =
+    black
+    mypy
+    ruff
+skip_install = true
+commands =
+    python setup.py check --strict --metadata
+    ruff check .
+    black --check --diff .
+    mypy --install-types --non-interactive .
```
Comment (stephrdev) [inline tox.ini:30]: Does it make sense to have mypy running if we don't do type annotations? Or do you intent to add them later?

**Review by stephrdev — COMMENTED** [review]
Thank you very much for the hard work you've put in my very old implementation for the Conrad relay cards. Appreciated!!

I added some notes and hints. Would be happy to learn about your thoughts on them.

Pleas don't forget to have yourself credited in the setup.py/pyproject.toml and/or AUTHORS file.

File: dev_requirements.txt (hunk 2 above)
Comment (eugenkleinschmidt) [reply dev_requirements.txt:1]: I've never worked with poetry till now. What about pip-tools, if it is only depencecies?

Comment (stephrdev) [reply dev_requirements.txt:1]: I'm also fine to keep it as it is right now as its not too complex.

File: src/relaycard/card.py (hunk 4 above)
Comment (eugenkleinschmidt) [reply src/relaycard/card.py]: Done. Have a look. I think most of the exceptions can be removed, when creating a proper types with ranges https://docs.python.org/3/library/typing.html#typing.Annotated

File: src/relaycard/frame.py (hunk 5 above)
Comment (eugenkleinschmidt) [reply src/relaycard/frame.py]: Done

File: tox.ini (hunk 6 above)
Comment (eugenkleinschmidt) [reply tox.ini:30]: Done

**Comment by stephrdev** [comment]
Wonderful. I assume you had it tested with a real relaycard, therefore I'll merge it now.

**Comment by eugenkleinschmidt** [comment]
> Wonderful. I assume you had it tested with a real relaycard, therefore I'll merge it now.

Yes, but CLI has some bugs. Only `--scan` and `--get-porst` are working. Maybe it could be updated to https://github.com/swansonk14/typed-argument-parser while bugfixing

**Comment by stephrdev** [comment]
I'd like to keep the number of external dependencies small, having pyserial is fine but adding more should be considered twice.

**Pull request merged**